#==========Outline of this script================================================================================

# The scRNA-seq data used in this analysis were generated by Ge et al (Ge et al.,2020 PNAS) and obtained from GEO (accession code GSE124901).
# In this analysis, we loaded GSM2558064 and GSM2558065 to construct Seurat objects for young and old samples, respectively.
# Data processing codes used in the original paper are available in GitHub (https://github.com/nyuhuyang/scRNAseq-MouseSkinEpithelia).

#====== 0  Setup environment===================================================================================================================
setwd("/R/supplementaryFig10")
install.packages("~/Data/Seurat_2.3.4.tar.gz", repos = NULL, type = "source")
library(Seurat)
library(MAST)
library(SingleR)
library(dplyr)
library(ggrepel)
library(gplots)
source("Seurat3_functions.R") # generated by Ge_etal 
source("Seurat_functions.R") # generated by Ge_etal 

#====== 1  Reconstruction of clustering in Ge et al., 2020 PNAS================================================================================

# We partially modified the R code and comments published by Ge et al as “Seurat_setup.R” in GitHub.

#======1.1 Preparation of Seurat object =======================================================================================================
# Load the dataset
# Setup Seurat objects. Since count matrices contain filtered cells, we do not perform any additional filtering here.
Young <- Read10X(data.dir = paste0("/Downloads/data/young/"))
Aged <- Read10X(data.dir = paste0("/Downloads/data/aged/"))

MouseSkin_raw <- list("young" = Young,
                      "old" = Aged)
MouseSkin_Seurat <- list()
samples <- c("young","old")
conditions <- c("young","old")

for(i in 1:length(samples)){
  colnames(MouseSkin_raw[[i]]) <- paste0(samples[i],"_",colnames(MouseSkin_raw[[i]]))
  MouseSkin_Seurat[[i]] <- CreateSeuratObject(MouseSkin_raw[[i]],
                                              min.cells = 3,
                                              min.genes = 200,
                                              names.delim = "_")
  MouseSkin_Seurat[[i]]@meta.data$conditions <- conditions[i]
}
MouseSkin_Seurat <- lapply(MouseSkin_Seurat, FilterCells, 
                           subset.names = "nGene", 
                           low.thresholds = 200, 
                           high.thresholds = Inf) %>% 
  lapply(NormalizeData) %>% 
  lapply(ScaleData) %>% 
  lapply(FindVariableGenes, do.plot = FALSE)


# We will take the union of the top 1k variable genes in each dataset for alignment. Note that we use 1k genes in the manuscript examples. You can try this here with negligible changes to the overall results
genes.use <- lapply(MouseSkin_Seurat, function(x) head(rownames(x@hvg.info), 1000))
genes.use <- unique(unlist(genes.use))
for(i in 1:length(samples)){
  genes.use <- intersect(genes.use, rownames(MouseSkin_Seurat[[i]]@scale.data))
}
length(genes.use) # 1/10 of total sample size

#======1.2 Perform a canonical correlation analysis (CCA) =====================================================================================
# Run a canonical correlation analysis to identify common sources of variation between the two datasets.

MouseSkin <- RunCCA(MouseSkin_Seurat[[1]], MouseSkin_Seurat[[2]], 
                    genes.use = genes.use,num.cc = 50)
save(MouseSkin, file = "./data/MouseSkin_alignment.Rda")
remove(MouseSkin_raw,MouseSkin_Seurat)

# Calculate median UMI per cell
MouseSkin_raw_data <- as.matrix(x = MouseSkin@raw.data)
MouseSkin_young <- MouseSkin_raw_data[,grepl("young_",colnames(MouseSkin_raw_data))]
MouseSkin_old <- MouseSkin_raw_data[,grepl("old_",colnames(MouseSkin_raw_data))]
mean(colSums(MouseSkin_young)); mean(colSums(MouseSkin_old))
median(colSums(MouseSkin_young)); median(colSums(MouseSkin_old))
par(mfrow = c(1,2))
boxplot(colSums(MouseSkin_young), ylim = c(0, 21000))
title(main = "nUMI per cell in young mouse skin sample")
boxplot(colSums(MouseSkin_old),ylim = c(0, 21000))
title(main = "nUMI per cell in old mouse skin sample")
remove(MouseSkin_raw_data,MouseSkin_young,MouseSkin_old)

# CCA plot CC1 versus CC2 and look at a violin plot
p1 <- DimPlot(object = MouseSkin, reduction.use = "cca", group.by = "conditions", 
              pt.size = 0.5, do.return = TRUE)
p2 <- VlnPlot(object = MouseSkin, features.plot = "CC1", group.by = "conditions", 
              do.return = TRUE)
plot_grid(p1, p2)

PrintDim(object = MouseSkin, reduction.type = "cca", dims.print = 1:2, genes.print = 10)

p3 <- MetageneBicorPlot(MouseSkin, grouping.var = "conditions", dims.eval = 1:50, 
                        display.progress = TRUE, smooth = TRUE) # run on cluster
p3 + geom_smooth(method = 'loess')

DimHeatmap(object = MouseSkin, reduction.type = "cca", cells.use = 500, dim.use = c(1:3,25:20), 
           do.balanced = TRUE)

#======1.3 QC =================================================================================================================================
MouseSkin <- CalcVarExpRatio(object = MouseSkin, reduction.type = "pca",
                             grouping.var = "orig.ident", dims.use = 1:20)
MouseSkin
MouseSkin <- SubsetData(MouseSkin, subset.name = "var.ratio.pca",accept.low = 0.5) #10291 out of 10350
MouseSkin

mito.genes <- grep(pattern = "^mt-", x = rownames(x = MouseSkin@data), value = TRUE)
percent.mito <- Matrix::colSums(MouseSkin@raw.data[mito.genes, ])/Matrix::colSums(MouseSkin@raw.data)
MouseSkin <- AddMetaData(object = MouseSkin, metadata = percent.mito, col.name = "percent.mito")
MouseSkin <- ScaleData(object = MouseSkin, genes.use = genes.use, display.progress = FALSE, 
                       vars.to.regress = "percent.mito")

# Now we can run a single integrated analysis on all cells
VlnPlot(object = MouseSkin, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 1)
par(mfrow = c(2, 1))
GenePlot(object = MouseSkin, gene1 = "nUMI", gene2 = "percent.mito", ylim = c(0, 0.6))
GenePlot(object = MouseSkin, gene1 = "nUMI", gene2 = "nGene", ylim = c(0, 4000))

MouseSkin <- FilterCells(object = MouseSkin, subset.names = c("nGene", "percent.mito"), 
                         low.thresholds = c(500, -Inf), high.thresholds = c(3500, 0.20))
GenePlot(object = MouseSkin, gene1 = "nUMI", gene2 = "percent.mito", 
         xlim = c(0,20000), ylim = c(0, 0.6))
GenePlot(object = MouseSkin, gene1 = "nUMI", gene2 = "nGene", 
         xlim = c(0,20000), ylim = c(0, 4000))

#======1.4 Align seurat objects ===============================================================================================================
# Now we align the CCA subspaces, which returns a new dimensional reduction called cca.aligned
set.seed(42)
MouseSkin <- AlignSubspace(object = MouseSkin, reduction.type = "cca", grouping.var = "conditions", 
                           dims.align = 1:20)
MouseSkin@project.name <- "Fuchs"
# Now we can run a single integrated analysis on all cells

MouseSkin <- FindClusters(object = MouseSkin, reduction.type = "cca.aligned", dims.use = 1:20, 
                          resolution = 1.0, force.recalc = T, save.SNN = TRUE)

MouseSkin <- RunTSNE(object = MouseSkin, reduction.use = "cca.aligned", dims.use = 1:20, 
                     do.fast = TRUE)

p1 <- TSNEPlot(MouseSkin, do.return = T, pt.size = 1, group.by = "orig.ident")
p2 <- TSNEPlot(MouseSkin, do.return = T, pt.size = 1, group.by = "ident")
plot_grid(p1, p2)

TSNEPlot(object = MouseSkin,do.label = T, group.by = "ident", 
         do.return = TRUE, no.legend = T,
         pt.size = 1,label.size = 8 )+
  ggtitle("TSNE plot of young and old mouse skin")+
  theme(text = element_text(size=20),     							
        plot.title = element_text(hjust = 0.5))

save(MouseSkin, paste0("data/MouseSkin_",gsub("-","",Sys.Date()),".Rda"))

#======1.5 Confirmation of clustering results===================================================================================================
# Our tSNE plot was similar but slightly different from tSNE plots in Fig.1A of Ge et al.,2020 PNAS.
# To confirm whether we obtained essentially consistent clustering results, we investigated marker genes in each cell cluster in both cases.
# Ge et al. used R shiny app to visualize gene expressions in their tSNE plots (https://weillcornellmed.shinyapps.io/MouseSkinEpithelia/).

# Identification of top10 marker genes in each cluster
# cluster0
cluster0.Markers = FindMarkers(MouseSkin, ident.1 = "0", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster0.Markers.top10 = rownames(head(cluster0.Markers, 10))
# cluster1
cluster1.Markers = FindMarkers(MouseSkin, ident.1 = "1", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster1.Markers.top10 = rownames(head(cluster1.Markers, 10))
# cluster2
cluster2.Markers = FindMarkers(MouseSkin, ident.1 = "2", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster2.Markers.top10 = rownames(head(cluster2.Markers, 10))
# cluster3
cluster3.Markers = FindMarkers(MouseSkin, ident.1 = "3", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster3.Markers.top10 = rownames(head(cluster3.Markers, 10))
# cluster4
cluster4.Markers = FindMarkers(MouseSkin, ident.1 = "4", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster4.Markers.top10 = rownames(head(cluster4.Markers, 10))
# cluster5
cluster5.Markers = FindMarkers(MouseSkin, ident.1 = "5", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster5.Markers.top10 = rownames(head(cluster5.Markers, 10))
# cluster6
cluster6.Markers = FindMarkers(MouseSkin, ident.1 = "6", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster6.Markers.top10 = rownames(head(cluster6.Markers, 10))
# cluster7
cluster7.Markers = FindMarkers(MouseSkin, ident.1 = "7", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster7.Markers.top10 = rownames(head(cluster7.Markers, 10))
# cluster8
cluster8.Markers = FindMarkers(MouseSkin, ident.1 = "8", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster8.Markers.top10 = rownames(head(cluster8.Markers, 10))
# cluster9
cluster9.Markers = FindMarkers(MouseSkin, ident.1 = "9", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster9.Markers.top10 = rownames(head(cluster9.Markers, 10))
# cluster10
cluster10.Markers = FindMarkers(MouseSkin, ident.1 = "10", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster10.Markers.top10 = rownames(head(cluster10.Markers, 10))
# cluster11
cluster11.Markers = FindMarkers(MouseSkin, ident.1 = "11", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster11.Markers.top10 = rownames(head(cluster11.Markers, 10))
# cluster12
cluster12.Markers = FindMarkers(MouseSkin, ident.1 = "12", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster12.Markers.top10 = rownames(head(cluster12.Markers, 10))
# cluster13
cluster13.Markers = FindMarkers(MouseSkin, ident.1 = "13", logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
cluster13.Markers.top10 = rownames(head(cluster13.Markers, 10))


#FeaturePlot for cluster markers
FeaturePlot(MouseSkin, features = cluster0.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster1.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster2.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster3.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster4.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster5.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster6.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster7.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster8.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster9.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster10.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster11.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster12.Markers.top10, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = cluster13.Markers.top10, cols.use = c("lightgray", "red"))

# We also visualized expression patterns of cluster_markers in Ge et al's tSNE plot.
# We concluded that our clustering results is essentially consistent to the original one.
# From this comparison, we annotated our cell clusters as follows;
# Our_ids   :: cluster name in Ge et al.,2020 PNAS
# cluster0 = Basal Epd
# cluster1 = Infund
# cluster2 = SupraBasal Epd
# cluster3 = Basal Epd
# cluster4 = Isthmus
# cluster5 = Bulge
# cluster6 = Proliferation
# cluster7 = Hair germ
# cluster8 = Basal Epd
# cluster9 = SupraBasal HF
# cluster10 = Basal Epd
# cluster11 = SebGland
# cluster12 = Bulge
# cluster13 = Basal Epd

#======1.6 Rename Ident========================================================================================================================

# This list of cluster name was the same to that used in Ge et al.
new.ident = c("Basal Epd", "Infund", "SupraBasal Epd", "Basal Epd", 
              "Isthmus", "Bulge", "Proliferation", "Hair germ", 
              "Basal Epd", "SupraBasal HF", "Basal Epd", "SebGland", 
              "Bulge", "Basal Epd")

for(i in 0:13){
  MouseSkin = RenameIdent(object = MouseSkin, old.ident.name = i, 
                          new.ident.name = new.ident[i + 1])
}

TSNEPlot(MouseSkin, do.label = TRUE, group.by = "ident", pt.size = 1, label.size = 8)

#======1.7 Confirmation of marker expressions in our clusters==================================================================================
# We made a gene list of cluster markers based on Fig. 1B in Ge_et al., 2020 PNAS.
markers_Bulge = c('Krt24', 'Cd34', 'Grem1', 'Id2', 'Tnc', 'Thsd1', 'S100a4', 'Sfrp1', 'Dkk3', 'Pdzrn4', 'Postn', 'Angptl7')
markers_HG = c('Cldn10', 'Lhx2', 'Lgr5', 'Tgm5', 'Sox9', 'Cd200', 'Krt17', 'Fst', 'Aqp3', 'Aldh3a1', 'Sostdc1', 'Gstm1', 'Gstm5', 'Calml3', 'Cdh3') # We add Cdh3. 
markers_Isthmus = c('Gli1', 'Lgr6', 'Lrig1', 'Plet1', 'Gata6')
markers_SG = c('Pparg', 'Prdm1', 'Igfbp2', 'Mgst1')
markers_SupraBasalHF = c('Sprr1a', 'Krt79', 'Defb6', 'Cst6', 'Alox12e')
markers_BasalEpd = c('Ly6a', 'Klf5', 'Tfap2c', 'Gata3', 'Pou3f1', 'Ifngr1', 'Il1r2', 'Il6ra', 'Il33', 'Ndufa4l2', 'Serpinb2')
markers_SupraBasalEpd = c('Grhl1', 'Krt10', 'Krt77', 'Mt4', 'Krt1', 'Krtdap', 'Calm4')
markers_Proliferation = c('Ccnb1', 'Ccnb2', 'Cdk1', 'Top2a', 'Cdc20', 'Mki67', 'Cdca3', 'Birc5', 'Cenpf', 'Nusap1', 'Ube2c')
markers.to.plot = c(markers_Bulge, markers_HG, markers_Isthmus, markers_SG, markers_SupraBasalHF, markers_BasalEpd, markers_SupraBasalEpd, markers_Proliferation)

# Feature plots
FeaturePlot(MouseSkin, features = markers_Bulge, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = markers_HG, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = markers_Isthmus, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = markers_SG, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = markers_SupraBasalHF, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = markers_BasalEpd, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = markers_SupraBasalEpd, cols.use = c("lightgray", "red"))
FeaturePlot(MouseSkin, features = markers_Proliferation, cols.use = c("lightgray", "red"))

# DotPlots
SplitDotPlotGG(MouseSkin,
               genes.plot = rev(markers.to.plot),
               cols.use = c("red", "blue"),
               x.lab.rot = TRUE,
               plot.legend = TRUE,
               grouping.var = "orig.ident")


#====== 2  Supplementary Fig. 10===============================================================================================================
#======2.0 Change identity class of cells==================================================================================================
# Save clustering result as "CellType" in meta.data
MouseSkin = StashIdent(object = MouseSkin, save.name = "CellType")

# Change active ident to "orig.ident"
MouseSkin = SetAllIdent(object = MouseSkin, id = "orig.ident")

# Combine cluster_id and orig,ident (CellType and orig.ident), and make new ident, "CellType_young_old".
MouseSkin <- CombineIdent(object = MouseSkin, attribute.1 = "CellType", attribute.2 = "orig.ident")
MouseSkin = StashIdent(object = MouseSkin, save.name = "CellType_young_old")

# Check active ident 
TSNEPlot(object = MouseSkin)

#======2.1 Examination of gene expression patterns of region-specific ECMs identified in our RNA-seq analysis (Supplementary Fig. 10 a-b)=====================================
# These region-specific ECM genes were identified and classified in Fig. 3 
clustered_gene_in_Epi = read.csv("clusteredgeneinepi.csv", header = TRUE)
group1 = clustered_gene_in_Epi$group1
group2 = clustered_gene_in_Epi$group2
group3 = clustered_gene_in_Epi$group3
group4 = clustered_gene_in_Epi$group4
group5 = clustered_gene_in_Epi$group5
group6 = clustered_gene_in_Epi$group6
group7 = clustered_gene_in_Epi$group7
group8 = clustered_gene_in_Epi$group8
group9 = clustered_gene_in_Epi$group9
group10 = clustered_gene_in_Epi$group10
region_specific_ECMs_epi = c(group1, group2, group3, group4, group5, group6, group7, group9, group10) 
region_specific_ECMs_epi = region_specific_ECMs_epi[region_specific_ECMs_epi>0] #remove blank

# Calculation of average gene expressions in each cluster in "CellType_young_old"
cluster.average = AverageExpression(object = MouseSkin, return.seurat = TRUE)

# Construction of heatmap source data
order = c("Basal Epd_young", "Basal Epd_old", "Infund_young", "Infund_old", "Isthmus_young", "Isthmus_old", "Bulge_young", "Bulge_old", "Hair germ_young", "Hair germ_old", "SebGland_young", "SebGland_old", "SupraBasal Epd_young", "SupraBasal Epd_old", "SupraBasal HF_young", "SupraBasal HF_old", "Proliferation_young", "Proliferation_old")
heatmap_all = as.matrix(cluster.average@data)
all.genes = rownames(heatmap_all)
heatmap_all = as.data.frame(heatmap_all) 
heatmap_all = heatmap_all %>%
  dplyr::filter(all.genes %in% region_specific_ECMs_epi) %>%
  dplyr::select(order)

# Extract cells of young samples
heatmap_young = heatmap_all %>% dplyr::select(ends_with("_young"))

# Scaling
heatmap_young_scale = scale(t(heatmap_young))
heatmap_young_scale = t(heatmap_young_scale)　%>% tidyr::replace_na(replace = 0)

# Heatmap (Supplementary Fig. 10a)
hd = dist(heatmap_young_scale)
hr = hclust(hd, method = "complete")
clusters = cutree(hr, 7)
clust.height = length(unique(as.vector(clusters)))
mycol = colorRampPalette(c("red", "purple3", "deepskyblue", "violet", "gray27", "darkgreen", "gray57"))
mycol.height = mycol(clust.height)

heatmap.2(as.matrix(heatmap_young_scale),
          main = "all clustered genes in epi",
          Colv = FALSE, dendrogram = "row", cexCol = 1, cexRow = 1,
          RowSideColors = mycol.height[clusters],
          col = bluered(64), trace = "none", 
          margins = c(10,5), 
          offsetCol = 0,offsetRow = 0, 
          key.title = NA, key.xlab = "z-scored average expression", 
          sepcolor = "black", colsep = 1:ncol(heatmap_young_scale), rowsep = 1:nrow(heatmap_young_scale), sepwidth = c(0.005, 0.005))

# Source data
write.csv(heatmap_young, "FigS10a_heatmap_source_data.csv")
write.csv(heatmap_young_scale, "FigS10a_heatmap_source_data_scale.csv")

# Search for ECM genes not included in scRNA-seq data (Supplementary Fig. 10b)
d = setdiff(region_specific_ECMs_epi, rownames(heatmap_all))
d
#"Emilin3", "Ntng2", "Ndnf", "Nov", "Lamc3", "Col3a1", "Lum", "Gldn", "Col15a1", "Dpt", "Fgl2", "Spon1", "Fbn1", "Col6a3", "Tnfaip6", "Slit2", "Vcan", "Spock1", "Rspo1"


#======2.2 Differential expression testing for HG_ECMs between "Hair germ_young" and "Hair germ_old" (Supplementary Fig. 10c)================
# HG_ECMs are defined as group 6, 7, 8 and 9 in Fig. 3
HG_ECMs = c(group6, group7, group8, group9)
HG_ECMs = HG_ECMs[HG_ECMs>0] # remove blank
HG_ECMs = HG_ECMs[-which(HG_ECMs %in% c("Ndnf", "Nov", "Lamc3", "Col3a1", "Lum", "Gldn", "Col15a1", "Dpt", "Fgl2", "Spon1", "Fbn1", "Col6a3", "Tnfaip6", "Slit2", "Vcan", "Spock1", "Rspo1"))] # Remove ECMs genes not detected in the analysis above.

# Differential expression testing 
# We used MAST; Model-based Analysis of Single-cell Transcriptomics (Finak et al.,2015)
MAST_HG = FindMarkers(MouseSkin, ident.1 = "Hair germ_old", ident.2 = "Hair germ_young", genes.use = HG_ECMs, test.use = "MAST", logfc.threshold = 0, min.pct = 0)

# Log2FC
avg_FC = expm1(MAST_HG$avg_logFC)+1
avg_log2FC = log2(avg_FC) 

# -Log10p_adj
log10p_val_adj = -log10(MAST_HG$p_val_adj)

# Volcano plot
vd = cbind(rownames(MAST_HG), as.numeric(avg_log2FC), as.numeric(log10p_val_adj))
vd = as.data.frame(vd)
vd= vd %>%
  dplyr::rename(gene = V1) %>%
  dplyr::mutate(class = ifelse(log10p_val_adj > -log10(0.05), "adj_p_value<0.05","others"))

label = vd$gene
selected_genes = c("Thbs1", "Abi3bp", "Fbln1", "Mfap2", "Ltbp4", "Lamc1", "Bgn", "Fras1", "Col4a1", "Hmcn1")
label[!label %in% selected_genes] = ""

ggplot(vd, aes(x = avg_log2FC, y = log10p_val_adj, colour = class, label = label)) +
  xlim(-1, 1) +
  xlab("Average log2 (fold change)") +
  ylab("-Log10 (adjusted p value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", colour = "gray35") +
  geom_vline(xintercept = 0, linetype = "dashed", colour = "gray35") +
  geom_point() +
  scale_color_manual(values = c("red", "black")) +
  geom_text_repel()

# Output source data
vd = vd %>% dplyr::rename(avg_log2FC = V2, log10_adj_p_val = V3)
write.csv(MAST_HG, "MASTDEtest_HG_genes.csv")
write.csv(vd, "FigS10c_volcanoplot.csv")

#======2.3 Differential expression testing for Bulge_ECMs between "Bulge_young" and "Bulge_old" (Supplementary Fig. 10d)=====================
# Bulge_ECMs are defined as group1, 2, 3, 4, 7 and 10 in Fig. 3
Bulge_ECMs = c(group1, group2, group3, group4, group7, group10)
Bulge_ECMs = Bulge_ECMs[Bulge_ECMs>0] # remove blank
Bulge_ECMs = Bulge_ECMs[-which(Bulge_ECMs %in% c("Emilin3", "Spock1"))] # remove not detected ECMs in scRNA-seq data

# Differential expressing testing with MAST
MAST_Bulge = FindMarkers(MouseSkin, ident.1 = "Bulge_old", ident.2 = "Bulge_young", genes.use = Bulge_ECMs, test.use = "MAST", logfc.threshold = 0, min.pct = 0)

# Log2FC
avg_FC = expm1(MAST_Bulge$avg_logFC)+1
avg_log2FC = log2(avg_FC) 

# -Log10_adj_p_val
log10p_val_adj = -log10(MAST_Bulge$p_val_adj)

# Volcano plot
vd = cbind(rownames(MAST_Bulge), as.numeric(avg_log2FC), as.numeric(log10p_val_adj))
vd = as.data.frame(vd)
vd= vd %>%
  dplyr::rename(gene = V1) %>%
  dplyr::mutate(class = ifelse(log10p_val_adj > -log10(0.05), "adj_p_value<0.05","others"))

label = vd$gene
selected_genes = c("Lamb3", "Vwa1", "Col8a2","Sparc", "Igfbp4", "Tgfbi", "Adamtsl4", "Col18a1", "Col17a1", "Vwa2", "Egfl6", "Ecm1", "Postn", "Col6a1", "Igfbp5", "Igfbp7", "Npnt")
label[!label %in% selected_genes] = ""

ggplot(vd, aes(x = avg_log2FC, y = log10p_val_adj, colour = class, label = label)) +
  xlim(-1, 1) +  
  xlab("Average log2 (fold change)") +
  ylab("-Log10 (adjusted p value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", colour = "gray35") +
  geom_vline(xintercept = 0, linetype = "dashed", colour = "gray35") +
  geom_point() +
  scale_color_manual(values = c("red", "black")) +
  geom_text_repel()

# Output source data
vd = vd %>% dplyr::rename(avg_log2FC = V2, log10_adj_p_val = V3)
write.csv(MAST_Bulge, "MASTDEtest_Bulge_genes.csv")
write.csv(vd, "FigS10d_volcanoplot.csv")


